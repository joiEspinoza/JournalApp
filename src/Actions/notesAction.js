import { db } from "../FireBase/fireBaseConfig";
import { HloadNotes } from "../Helpers/HloadNotes";
import types from "../Types/types";
import Swal from "sweetalert2";
import { HFileUpload } from "../Helpers/HFileUpload";

/////

const starNewNote = () =>
{
    return async ( dispatch, getState ) => {

        try 
        {
            const { uid } = getState().auth;

            const newNote = 
            {
                title : "",
                body : "",
                date : new Date().getTime()
            };
    
            const docRef = await db.collection( `${ uid }/journal/notes` ).add( newNote );
    
            dispatch( activeNote( docRef.id, newNote ) );
            
            dispatch( addNewNote( docRef.id, newNote ) );
        } 
        catch (error) 
        {
            console.log(error);
        }
      

    };
};

/////

const activeNote = ( id, note ) =>( { type : types.notesActive, payload : { id, ...note } } );

///////

const addNewNote = ( id, note ) => ( { type : types.notesAddNew, payload : { id, ...note } } );

//////

const startLoadingNotes = ( uid ) =>
{
    return async ( dispatch ) => 
    {
        try 
        {
            const notes = await HloadNotes( uid );

            dispatch( setNotes( notes ) );
        } 
        catch (error) 
        {
            console.log( error );
        }
     
    };
};

///////

const setNotes = ( notes ) => ( { type : types.notesLoad, payload : notes } );

//////

const saveNote = ( note ) =>
{
    return async ( dispatch, getState ) =>
    {
        try 
        {
            const { uid } = getState().auth;

            if( !note.url )
            {
                delete note.url;
            };
    
            const noteToFireStore = { ...note };
            delete noteToFireStore.id; // elimina una propiedad del objeto
    
            await db.doc( `${ uid }/journal/notes/${ note.id }` ).update( noteToFireStore );
    
            dispatch( refreshNote( note.id, noteToFireStore ) );
    
            Swal.fire("Saved",note.title,"success");

        } 
        catch ( error ) 
        {
            Swal.fire("Error",error,"error");
        }
      
    };
};

//////

const refreshNote = ( id, note ) => ( { type : types.notesUpdated, payload : { id, note : { id, ...note } } } );

///////

const startUploadingPicture = ( file ) =>
{
    return async ( dispatch, getState ) =>
    {

        try 
        {
            const { active:activeNote } = getState().notes;

            Swal.fire(
    
                {
                    title: "Uploading....",
                    text: "Please wait...",
                    showConfirmButton: false,
                    onBeforeOpen: ()=>{
                        
                        Swal.showLoading();
                    }
                }
    
            );
    
            const fileUrl = await HFileUpload( file );
            activeNote.url = fileUrl;
            
            dispatch( saveNote( activeNote ) );
    
            Swal.close();
        } 
        catch (error) 
        {
            console.log( error );
        }
       
    };
};

//////

const startDeleteNote = ( id ) =>
{
    return async ( dispatch, getState ) =>
    {
        try 
        {
            const uid = getState().auth.uid;

            await db.doc( `${ uid }/journal/notes/${ id }` ).delete();
    
            dispatch( deleteNote( id ) );
        } 
        catch (error) 
        {
            console.log( error );
        }
       
    };
};

/////

const deleteNote = ( id ) => ( { type : types.notesDelete, payload : id } );

/////

const purgeLogout = () => ( { type : types.notesLogoutCleaning } );


/////

export 
{ 
    starNewNote, 
    activeNote, 
    setNotes, 
    startLoadingNotes, 
    saveNote, 
    refreshNote, 
    startUploadingPicture,
    startDeleteNote,
    purgeLogout
};

