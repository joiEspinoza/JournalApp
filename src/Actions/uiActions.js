import types from "../Types/types";

///////

const setErrorAction = ( error ) => ( {  type :  types.uiSetError, payload : error } );

const uiRemoveErrorAction = () => ( { type : types.uiRemoveError } );

const uiStartLoadingAction = () => ( { type : types.uiStartLoading, payload : true } );

const uiFinishLoading = () => ( { type : types.uiFinishLoading, payload : false } );


///////

export { setErrorAction,uiRemoveErrorAction, uiStartLoadingAction, uiFinishLoading };