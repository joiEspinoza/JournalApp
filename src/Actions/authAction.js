import types from "../Types/types";
import { firebase, googleAuthProvider } from "../FireBase/fireBaseConfig";
import { uiStartLoadingAction, uiFinishLoading } from "./uiActions";
import Swal from 'sweetalert2';
import { purgeLogout } from "./notesAction";

////

const login = ( uid, displayName ) => ( { type : types.login, payload : { uid, displayName } } );

///// LOGIN NORMAL

const starLoginEmailPassword = ( email, password ) =>
{
    return ( dispatch ) => 
    {
        dispatch( uiStartLoadingAction() );

        return firebase.auth().signInWithEmailAndPassword( email, password )
        .then( ( { user } ) => 
        { 
            dispatch( login( user.uid, user.displayName ) );

            dispatch( uiFinishLoading() );

        } )
        .catch( ( error ) => 
        { 
            console.log( error );
            
            dispatch( uiFinishLoading() );

            Swal.fire( "error", error.message, "error" );
            
        } );
    };
};

///// CREAR USUARIO

const starRegisterWhitEmailNamePassword = ( email, password, name ) =>
{
    return( ( dispatch ) => 
    {

        firebase.auth().createUserWithEmailAndPassword( email, password )
        .then( async ( { user } ) =>
        {  
            await user.updateProfile( { displayName : name } ); 
            dispatch( login( user.uid, user.displayName ) );

        } )
        .catch( ( error ) => {
            
            console.log( error );

            Swal.fire( "error", error.message, "error" );
        
        } );

    });
};

///// LOGIN GOOGLE

const startGoogleLogin = () =>
{
    return ( dispatch ) => 
    {
        firebase.auth().signInWithPopup( googleAuthProvider )
        .then( ( { user } ) => dispatch( login( user.uid, user.displayName ) ) )
        .catch( ( error ) => console.log( error ) );
    };
};

/////////////////////////////////////////////----LOGOUT----///////////////////////////////////////

const startLogout = () =>
{
    return async ( dispatch ) => 
    {
        try 
        {
            await firebase.auth().signOut();

            dispatch( purgeLogout() );
            dispatch( logout() );

        } 
        catch( error ) 
        {
            console.log( error );
        };
       
    };
};

const logout = () => ( { type : types.logout } );

//////

export 
{ 
    login, 
    starLoginEmailPassword, 
    startGoogleLogin, 
    starRegisterWhitEmailNamePassword,
    startLogout,
    logout 
};