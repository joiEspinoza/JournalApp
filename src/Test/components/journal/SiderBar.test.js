const { mount } = require("enzyme");
const { Provider } = require("react-redux");
const { default: SiderBar } = require("../../../Components/journal/SiderBar");
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from "redux-thunk";
import { startLogout } from '../../../Actions/authAction';
import { starNewNote } from '../../../Actions/notesAction';

///////

jest.mock( '../../../Actions/authAction', () => ( { 
    
    startLogout : jest.fn(), 

} ) );

jest.mock( '../../../Actions/notesAction', () => ( { 
    
    starNewNote : jest.fn(), 

} ) );

///////

const middlewares = [ thunk ]
const mockStore = configureStore( middlewares );

////////

const initState = 
{ 
    auth : { uid : "ABCD1234", name : "TestMan" }, 
    ui: { loading : false, msgError : null }, 
    notes : { note : [], active : null } 
};

let store = mockStore( initState );
store.dispatch = jest.fn();

////////

describe('Pruebas en SiderBar.js', () => 
{

    const wrapper = mount( 
    
        <Provider store={ store } >
            <SiderBar/> 
        </Provider>
    
    );

    ///////

    test('Debe mostrarse correctamente ', () => 
    {
        expect( wrapper ).toMatchSnapshot();
    });
    
    //////

    test('debe llamar logout', () => 
    {
        wrapper.find( ".btn" ).prop( "onClick" )()

        expect( startLogout ).toHaveBeenCalled();

    });
    
    //////

    test('debe llamar starNewNote ', () => 
    {
        wrapper.find( ".fa-calendar-plus" ).prop( "onClick" )();

        expect( starNewNote  ).toHaveBeenCalled();
        
    });
    
        
});
