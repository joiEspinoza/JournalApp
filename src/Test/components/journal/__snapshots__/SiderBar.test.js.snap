// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SiderBar.js Debe mostrarse correctamente  1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
        style={
          Object {
            "color": "white",
          }
        }
      />
      <span
        style={
          Object {
            "color": "white",
          }
        }
      >
         
        TestMan
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
  >
    <i
      className="far fa-calendar-plus fa-5x"
      onClick={[Function]}
    />
    <p
      className="mt-5"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
